11 фз

Maks1-15 opened the pull request:
Реализован функционал 11 фз

Maks1-15 committed:
11 фз

Реализован функционал 11 фз

Maks1-15 committed:
11 фз

Реализован функционал 11 фз

Maks1-15 committed:
11 фз

Реализован функционал 11 фз
Что такое безумие? Это когда из раза в раз делаешь

Maks1-15 committed:
11 фз

Реализован функционал 11 фз
Что такое безумие? Это когда из раза в раз делаешь push в надежде, что codestyle пройдет

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/UserRepositoryImpl.java:
```diff
@@ -7,21 +7,57 @@
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 @Repository
```
Это все таки `@Component `а не репозиторий

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/FilmRepositoryImpl.java:
```diff
@@ -1,27 +1,52 @@
     private final Map<Long, Film> films = new HashMap<>();
-    private final AtomicInteger idGenerator = new AtomicInteger(1);
+    private final AtomicLong idGenerator = new AtomicLong(1);
+    private final Map<Long, Set<Long>> filmLikes = new HashMap<>();
```
Создавать отдельно хранилище для лайков избыточно было проще добавить поле в Film
private Set<Long> userIds = new HashSet<>();
также и для user, лучше добавить отдельное поле friends

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/FilmRepositoryImpl.java:
```diff
@@ -1,27 +1,52 @@
+import java.util.concurrent.atomic.AtomicLong;
 
 @Repository
+@RequiredArgsConstructor
```
Данная аннотация избыточна, т.к. нет полей для инициализации в конструкторе

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/FilmRepositoryImpl.java:
```diff
@@ -1,27 +1,52 @@
+
+    // Метод для добавления лайка фильму от пользователя
+    public void addLike(Long filmId, Long userId) {
+        validIdByMap(filmId);
```
Проверку наличия фильма нужно делать в сервисном классе и при удалении лайка также

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/UserRepositoryImpl.java:
```diff
@@ -7,21 +7,57 @@
+
+    // Метод для добавления пользователя в друзья к другому пользователю
+    public void addFriend(Long userId, Long friendId) {
+        checkUsersExists(userId, friendId);
```
Проверку наличия пользователей, также нужно делать в сервисном классе

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/impl/FilmServiceImpl.java:
```diff
@@ -28,4 +32,24 @@ public Film update(Film film) {
+    }
+
+    public List<Film> getTopNFilms(int count) {
+        if (count < 0) {
```
Меньше или равно 0 
count <= 0

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/impl/UserServiceImpl.java:
```diff
@@ -31,4 +32,26 @@ public User update(User user) {
+    public List<User> getMutualFriends(Long userId, Long otherUserId) {
+        List<Long> mutualFriendIds = userRepository.getMutualFriends(userId, otherUserId);
+        return userRepository.getAll().stream()
+                .filter(user -> mutualFriendIds.contains(user.getId()))
```
Оптимально написать метод получения по id, вместо получения всех записей в хранилище и затем по ним проходить
mutualFriendIds.stream()
                .map(id -> userRepository.getByID(id))
                .collect(Collectors.toList());

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/impl/UserServiceImpl.java:
```diff
@@ -31,4 +32,26 @@ public User update(User user) {
+    public List<User> getFriends(Long userId) {
+        List<Long> friendIds = userRepository.getFriends(userId).stream().toList();
+        return userRepository.getAll().stream()
+                .filter(user -> friendIds.contains(user.getId()))
```
Здесь также не оптимально проходить по всем записям, у вас есть список id, напишите метод получения записи по id и дальше через steam + map преобразуйте

kesch9 reviewed (commented):
Несколько комментариев к работе

Maks1-15 committed:
11 фз

Внедрение новых правок
Изменение логики проверки данных (внесены в controller)
Упрощен код

kesch9 reviewed (approved):
Approve

Maks1-15 committed:
Промежуточное фз:
Добавление er-диаграммы базы данных
Добавлено описание и ссылка на источник.

a-dbr commented:
По промежуточному: на схеме не везде читаются названия столбцов. по таблице mpa_ratings: там столбец "продолжительность" вроде лишний. Добавь схему в ридми файл. В целом схема хорошая.

Maks1-15 committed:
Промежуточное фз:
Изменена диаграмма согласно правкам

a-dbr commented:
Отлично!
